### ISSUE TITLE
Allowing cycling tasks to run out of order.

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc#1538]
In a cycling suite, task proxies spawn their next instance (at the next cycle point) when they enter the 'submit' state, so even if a task fails its successor will still be ready for action.    However, the successors of any tasks downstream of the failed one will not be available, because the instance that depends on the failed task will not trigger, and thus will not spawn.

E.g. consider a clock-triggered task that retrieves obs, followed by another tasks that processes them:
```
[[[PT6H]]]
    graph = get_obs => proc_obs
```

If ```get_obs``` fails at cycle point T, obs retrieval will still run in subsequent cycles, but obs processing won't.

It's easy to get around this manually:
 * GUI: right-click on the waiting ```proc_obs[T]``` and choose "Force spawn".
 * CLI: ```cylc reset --state=spawn SUITE proc_obs T``` (this is not intuitive, we should probably have a ```cylc spawn``` command - subject to #1249)

It's also possible to get around this automatically, e.g.:
 ```
    graph = """get_obs => proc_obs & !spawner
                     get_obs:fail => spawner"""
```
where ```spawner``` uses the CLI to force-spawn ```proc_obs``` (and potentially to retrigger get_obs and reset itself to waiting as well, to effect a full retry):
```sh
[runtime]
   [[spawner]]
       script = """
# Spawn proc_obs successor if it hasn't already spawned.
cylc reset --state=spawn $CYLC_SUITE_NAME proc_obs $CYLC_TASK_CYCLE_POINT
# Make get_obs retry as well (note get_obs should do it's own retries before failing too).
cylc trigger $CYLC_SUITE_NAME get_obs $CYLC_TASK_CYCLE_POINT
# Reset myself to wait on the re-triggered get_obs.
cylc reset --state=waiting $CYLC_SUITE_NAME spawner $CYLC_TASK_CYCLE_POINT"""
```

However, it would be nice if cylc could just do the right thing here, with no intervention or extra tasks required.

**Comment by benfitzpatrick** [comment]
I have been thinking for a while that the 'active cycle points' tracking should be the generator of tasks, not the submit-successor mechanism. The change in behaviour would break some suites though, I think - those that implicitly rely on `foo[-PT6H]:submit => foo`.

**Comment by arjclark** [comment]
> The change in behaviour would break some suites though, I think - those that implicitly rely on `foo[-PT6H]:submit => foo`.

Agreed, that one's bound to catch people out.

**Comment by hjoliver** [comment]
Spawn-on-submit is often relied on at start-up too, e.g.:
```
[scheduling]
    [[dependencies]]
         [[[R1]]]
                graph = install => foo
         [[[P1D]]]
                graph = foo
```
Here, we only need the first instance of ```foo``` to depend on ```install```, because the second instance won't be created until the first submits, after ```install``` has finished.  Without the slightly delayed spawning (or some new mechanism with a similar effect) we'd need an artificial inter-cycle dependence on ```foo```, or else have all instances of ```foo``` depend explicitly on ```install``` at the first cycle point.  Strictly speaking  explicit dependence on ```install``` is the right thing to do, but it has other consequences (e.g. ```install``` will stick around forever).

**Comment by matthewrmshin** [comment]
> I have been thinking for a while that the 'active cycle points' tracking should be the generator of tasks, not the submit-successor mechanism.

I also believe that this is the most important point here. The task pool should be responsible to generate new task proxies.

**Comment by hjoliver** [comment]
(Historical reason for "spawn on submit": in the early days I thought different task proxy sub-classes might need different spawning behaviour.  E.g. "sequential" tasks were originally implemented by "spawn on succeed")

**Comment by hjoliver** [comment]
Even if no failure occurs, spawn-on-submit means that task X[T2] cannot run before X[T1] where T2 > T1 even X[T2]'s prerequisites are satisfied before then.  This sort of scenario is unlikely to occur normally, but it's possible. 

> The task pool should be responsible to generate new task proxies.

So, yes, I agree with this.

**Comment by hjoliver** [comment]
Some more historical context for spawn-on-submit (just reminding myself).  Prior to cylc-3 there was no suite definition, just a collection of "task definitions" which were templates for task proxy objects that self-organize by matching prerequisites and outputs.  Task proxies knew their own prerequisites and outputs (etc.), *and their own cycling sequences* (e.g. I should run at 06 and 18 hours daily) - so it seemed fairly obvious that they should be responsible for spawning their own successors (a task proxy knows the cycle point of its next instance).  That being the case, job submit time is the only self-spawning point that allows successive instances of a task to run in parallel *and* avoids - without any global limiting mechanism - the generation of multiple waiting instances of every task.  To this day the only downside of this that I'm aware of, in terms of limiting optimal scheduling across multiple interleaved cycles, is the topic of this issue.

Another way of looking at this problem is: the self-spawn-on-submit mechanism for generating new task proxies allows successive instances of a cycling task to run in order, or to overlap, or to run entirely in parallel, *but not to run out of order*.

**Comment by arjclark** [comment]
> ...not to run out of order... 

Potentially this could relate to the priority with which things get processed in #1196...

**Comment by hjoliver** [comment]
A clearer explanation of this in light of recent comms with @arjclark: spawn-on-submit effectively makes every cycling task sequential with respect to job submission (not execution).   Run the following example in gcylc dot view to clearly see (a) sequential job submission of tasks `foo` and `bar`, and (b) when `foo[5]` finishes before previous `foo` instances, `bar[5]` can't trigger immediately - when ideally it should - because it does not exist yet.
```
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    final cycle point = 10
    max active cycle points = 10
    [[dependencies]]
	[[[R//P1]]]
	   graph = foo => bar
[runtime]
    [[root]]
	script = sleep 10
    [[foo]]
	script = """
if [[ $CYLC_TASK_CYCLE_POINT == 5 ]]; then
   sleep 1
else
   sleep 10
fi"""
```

**Comment by hjoliver** [comment]
An idea for fixing this:

As discussed above, task proxies should no longer be responsible for spawning their own successors.  The earliest task state that can be triggered off is 'submitted', so we need task proxies to exist in the 'waiting' state at the time upstream depended-on tasks submit, regardless of the state of previous (earlier cycle point) instances of their own type.    So: *whenever a task proxy goes to the submitted state, the pool could generate a waiting instance of any dependent task proxy (as determined by the graph), if it doesn't already exist.*  @arjclark - can you see any problems with this?   One interesting consequence would be that re-triggering a finished task would naturally kick off its downstream workflow again too.  Sometimes this would be a good thing, but we'd need an option to choose when manually triggering a task.

**Comment by hjoliver** [comment]
If the above suggestion works out, we could take the opportunity to do this too: https://github.com/cylc/cylc/issues/108#issuecomment-9066707

**Comment by hjoliver** [comment]
(@arjclark - note I deleted a new comment that referenced you - had a better idea...)

**Comment by hjoliver** [comment]
I think I've solved this problem, just some detail to iron out...

**Comment by hjoliver** [comment]
**What should govern the creation of new task proxies?**

Firstly I'll address https://github.com/cylc/cylc/issues/1538#issuecomment-123259863 above from Ben and Matt:
>> I have been thinking for a while that the 'active cycle points' tracking should be the generator of tasks, not the submit-successor mechanism.

> I also believe that this is the most important point here.

Do you mean to instantiate all task proxies out to max-active-cycles??   That would work, but it would have serious performance implications for large suites with a lot of runahead, because task proxies can be quite heavyweight objects. (although this could change if in the future we delay loading `[runtime]` config until job submission time.)

That being the case, I think the goal is to have the minimum number of 'waiting' proxies in the pool without causing  delays due to them not existing yet when needed.

In fact, that is exactly what the current mechanism is designed to do (note that *"spawn next instance on submit"* is equivalent to *"retain exactly one 'waiting' instance of each task proxy"*). There is only one problem with this: one 'waiting' instance of a task is not enough IF it happens to be waiting on an upstream failed task, because it won't have spawned a successor to wait on subsequent instances of the failed task (which may succeed).   However, we can't simply up the number of waiting tasks of each type - there could be two failures, or three, and so on... 

**Proposed Solution**

Short of instantiating everything out to max-active-cycles I can only think of one way to handle this easily:

 1. The graph knows which tasks proxies are needed next.  The earliest task state we can trigger off is 'submitted', so when a tasks submits use the graph to determine its immediate downstream dependents and instantiate them if they have not already been instantiated.

 2. We still need spawn-next-instance-on-submit as well, in order to retain one waiting instance of any task that has no triggers; these won't be generated by 1.

I have tested this and it works (just need to deal with ~18 failing tests, which is expected as this adds some waiting tasks to the state dump and run-db).

@cylc/core - let me know if you disagree.

**Comment by hjoliver** [comment]
A minor consequence of the above proposed solution: previously, task proxies remembered if they had spawned already (in order to avoid doing it again).  Now, because tasks can potentially run out of order with respect to cycle point, the task pool will need to remember what task proxies have already been generated (in order to avoid generating them again).

**Comment by benfitzpatrick** [comment]
I still like instantiating out to max active cycles, and (as you suggest) countering the performance aspects by having lighter-weight task proxies.

**Comment by matthewrmshin** [comment]
It sounds like we have several considerations here.

1. I agree that the task pool should manage spawning. It is the correct object with a global view of all current task proxies, and the knowledge of how much resource they are using.
2. Perhaps we should address the task proxy memory issue first to give us more flexibility for solving the spawning issue?
3. There are 2 conflicting requirements here. On the one hand, we want to present as much of the suite graph as possible to the user, so they can see what is going on. On the other hand, the task pool should really only spawn a task proxy if it has a chance to run *soon*, e.g. if most of its dependencies are currently submitted or running. In the ideal world, there is no reason why we need to mix the 2 requirements together, i.e. the presentation of the suite graph should only require some task names, and their suite definitions, it should not require the full run time task proxy objects. If we can some how separate the solutions of these requirements, we should end up with a system that is more efficient. * Disclaimer: Not sure if I am talking gibberish or not.

**Comment by hjoliver** [comment]
Ben & Matt - I think you are pushing the longer term goals here, whereas I'm proposing a quick fix to the only outright bug identified in the current task generation mechanism.  These two things can be done sequentially of  course.  Let me be more explicit:

 1. I agree that using max-active-cycles as the constraint on task proxy creation would work and it is nice and simple.  **BUT** to avoid memory blowout it would require delayed loading of `[runtime]' config into task proxies.  I can think of several ways to do this, but it is probably a medium term goal (with non-trivial consequences, such as limiting display of all the extra waiting tasks in the gui).

2.  I also agree - as discussed elsewhere - that we ultimately need a way to limit the task pool size regardless of the total number of tasks defined in the suite.  Max active cycles only helps down to the single cycle minimum, which would not be sufficient in huge suites.  **BUT** how to do this (without stalling the suite...) is an unsolved problem, so this is probably a longer term goal.

3. Under the current regime (no global pool limiting, and no delayed runtime loading) the best we can do is minimize the number of waiting tasks.  As discussed above, spawn-on-submit implies exactly one waiting proxy of each type, which is close to minimal, but it has one deficiency - it doesn't let tasks run out of order.  I am proposing a relatively simple fix for this problem: when a task submits (i.e. the first moment at which other tasks can trigger off it), create its downstream dependents IF spawn-next-instance-on-submit hasn't already done so.  

If this is simple enough to do, and it fixes real problems (relayed to me by @arjclark recently, e.g. in MO post processing suites) is there any reason not to do it, at least as an interim solution?

**Comment by hjoliver** [comment]
Regarding Matt's 3rd point above, "presenting as much of the suite graph as possible to the user" has not been a concern of mine at all.  Retaining a single waiting instance of each task type is just the simplest way (with the out-of-order proviso above!) to ensure that any task proxy will be available when needed, without having vastly more waiting tasks than are needed.

**Comment by hjoliver** [comment]
> I agree that the task pool should manage spawning. 

Unfortunately the terminology "spawn on submit" gives the impression that task proxies are managing the generation of new instances themselves, rather than the task pool, which is not really true.  It is simply that - due to memory use and scheduling efficiency concerns - we try to keep only one waiting instance of each task proxy in the pool at all times.  I.e. our task pool limiting mechanism necessarily depends on the state of the individual tasks (and moreover it is the task proxies that know how to figure out the cycle point of their next instance, which I think is reasonable for "cycling tasks").  Under these circumstances there is no material difference between "the pool generates new proxies after asking each task their state and next cycle point" and "the pool tells task proxies to generate their next instance (under the same conditions)".

**Comment by hjoliver** [comment]
> Perhaps we should address the task proxy memory issue first to give us more flexibility for solving the spawning issue?

See #1689

**Comment by trwhitcomb** [comment]
Would it be possible in the shorter term to have a suite-level or system-level option that controls how many cycles in advance we would spawn a successor?

I'm starting to play around with satellite data now, and the issue @arjclark demonstrated earlier is definitely something that we'll run into - I've defined an externally-triggered workflow with no inter-cycle dependence (so I'm using cylc as a workflow manager without doing any cycling), and I've found that in some cases the first task finishes faster for some datasets then others, but I'm still limited by whichever cycle is currently running the initial task, which doesn't help with throughput.

I don't know if there's anything I can do to help, but wanted to revive this as it's a real issue I'm currently running into!

**Comment by hjoliver** [comment]
[meeting]
  * longer term, this is solved by #987 
  * when it goes in (cylc-8?) we need to provide a back-compat mode for users relying on current behaviour at start-up?
  * @trwhitcomb's suggestion above can be implemented as a short term optional solution
